src/libcore/ffi.rs

```diff
@@ -0,0 +1,94 @@
+// Copyright 2018 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+//! Implementation of a `va_list`.
+
+#[cfg(any(target_arch = "arm", target_arch = "mips", target_arch = "powerpc64",
+          target_arch = "x86", windows))]
+#[cfg(not(stage0))]
+/// Basic byte pointer variant of a `va_list`.
+pub type VaListImpl = *const i8;
+
+#[cfg(all(target_arch = "aarch64", not(windows)))]
+#[cfg(not(stage0))]
+#[repr(C)]
+#[derive(Debug)]
+/// AArch64 ABI implementation of a `va_list`. See the
+/// [Aarch64 Procedure Call Standard] for more details.
+///
+/// [AArch64 Procedure Call Standard]:
+/// http://infocenter.arm.com/help/topic/com.arm.doc.ihi0055b/IHI0055B_aapcs64.pdf
+pub struct VaListImpl {
+    stack: *mut (),
+    gr_top: *mut (),
+    vr_top: *mut (),
+    gr_offs: i32,
+    vr_offs: i32,
+}
+
+#[cfg(all(target_arch = "powerpc", not(windows)))]
+#[cfg(not(stage0))]
+#[repr(C)]
+#[derive(Debug)]
+/// PowerPC ABI implementation of a `va_list`.
+pub struct VaListImpl {
+    gpr: u8,
+    fpr: u8,
+    reserved: u16,
+    overflow_arg_area: *mut (),
+    reg_save_area: *mut (),
+}
+
+#[cfg(all(target_arch = "x86_64", not(windows)))]
+#[cfg(not(stage0))]
+#[repr(C)]
+#[derive(Debug)]
+/// x86_64 ABI implementation of a `va_list`.
+struct VaListImpl {
+    gp_offset: i32,
+    fp_offset: i32,
+    overflow_arg_area: *mut (),
+    reg_save_area: *mut (),
+}
+
+/// A wrapper for a `va_list`
+#[cfg(not(stage0))]
+#[lang = "va_list"]
+#[unstable(feature = "c_variadic",
+           reason = "the `c_variadic` feature has not been properly tested on \
+                     all supported platforms",
+           issue = "27745")]
+#[derive(Clone, Copy, Debug)]
+// FIXME(dlrobertson): Do not use a *mut here. Fix core::intrinsics and
+// rustc_typeck::check::intrinsic::check_intrinsic_type.
+pub struct VaList(*mut VaListImpl);
```

Comment:
I'd really like this to be a reference instead of a pointer. I'm still working on trying to figure out how that can be implemented in `core::intrinsics` and `rustc_typeck::check::intrinsic::check_intrinsic_type`.

Reply:
This type is still missing `#[repr(transparent)]` FWIW. Also, a reference "should just work", if it causes issues we should discuss them, should be solvable.

Reply:
> This type is still missing #[repr(transparent)]

Ugh, don't know how I missed that. Thanks

> a reference "should just work"

I think something is wrong with my implementation of `check_intrinsic_type`. I get the following error.

```
error[E0308]: intrinsic has wrong type
    --> libcore/intrinsics.rs:1423:5
     |
1423 |     pub fn va_arg<'a, T: Copy>(ap: VaList<'a>) -> T;
     |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected bound lifetime parameter 'a, found concrete lifetime
     |
     = note: expected type `for<'a> unsafe extern "rust-intrinsic" fn(ffi::VaList<'a>) -> T`
                found type `unsafe extern "rust-intrinsic" fn(ffi::VaList<'a>) -> P0`
```

Reply:
Ah yeah, cc @nikomatsakis. I'm not sure if other intrinsics need lifetimes, if that's the case you can find a workable example there.

Reply:
The other intrinsics that use lifetimes use *late-bound* lifetimes -- i.e., lifetimes that only appear in the arguments/return-type. I guess that should work here too? I can't really tell what this error comes from -- is it from this branch, or something else?

Reply:
I temporarily removed the reference in favor of a `*mut` so I could get stuff to compile and get some reviews on some other stuff. I'll try push my WIP updates shortly.

Reply:
Updated

Reply:
I just noticed [this](https://github.com/rust-lang/rust/blob/master/src/libcore/intrinsics.rs#L55-L56) comment. Should the inner value of the `VaList` be a raw pointer?

Reply:
@nikomatsakis any input on this? I added the lifetime back.